fix(tianmu): fix query syntax (WHERE NOT IN ) is not supported. (#767)

lujiashun opened the pull request:
<!--

Thank you for contributing to StoneDB!

PR Title Format: <type>(<scope>): description to this pr (#issue_id)
e.g.
fix(util): fix sth..... (#1)

-->

## Summary about this PR
<!--

There MUST be one line starting with "Issue Number:  " and 
linking the relevant issues via the "close" or "ref".
e.g.:
Issue: close #1

-->

Issue Number: ref #767

[summary]
1 'sub_select' item is the left argument,not the right argument of 'item_func';
2 5.7 use 'item_func_not' class to represent the meaning of the syntax "not", while 8.0 use a variable 'value_transform';
3 porting new join code from 5.7 for this case；
3.1 solve throwing exception  ”select * from t1 where val not in (select * from t2 where val2 >= 10);“
4 porting ChooseJoinAlgorithm from 5.7 for this case;
4.1 solve incorrect result set for some sql in the mtr（which one is forgotton...)
5 porting ResetToTemplate modification from 5.7 for this case;
5.1 solve incorrect result set "select * from t1 where val not in (select * from t2 where val2 > t1.val);"


## Tests Check List
<!-- At least one of next options must be included. -->

- [ ] Unit test
- [X] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

## Changelog
<!-- At least one of next options must be included. -->

- [ ] New Feature
- [X] Bug Fix
- [ ] Improvement
- [ ] Performance Improvement
- [ ] Build/Testing/CI/CD
- [ ] Documentation
- [ ] Not for changelog (changelog entry is not required)

## Documentation
<!-- At least one of next options must be included. -->

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features

RingsC commented on storage/tianmu/core/query.cpp:
```diff
@@ -448,9 +448,16 @@ void Query::ExtractOperatorType(Item *&conds, common::Operator &op, bool &negati
+    case Item_func::UNKNOWN_FUNC: {
       op = common::Operator::O_UNKNOWN_FUNC;
+      if (cond_func->arg_count == 2) {
+        auto subselect_item = dynamic_cast<Item_in_subselect *>(cond_func->arguments()[1]);
```
Here, we can use `down_cast`, which implemented by MySQL.

RingsC replied:
Ok, Here we canot sure which type of this args. And, the cast maybe failed. Therefore, keep this.

RingsC commented on storage/tianmu/core/query.cpp:
```diff
@@ -1685,6 +1693,9 @@ bool Query::ClearSubselectTransformation(common::Operator &oper_for_subselect, I
     field_for_subselect = nullptr;
+    // the left side of equality should be the left side of the original
+    // expression with subselect
+    left_ref = ((Item_func *)cond_removed)->arguments()[0];
```
Using c++ style type casting.

RingsC commented on storage/tianmu/core/query_compile.cpp:
```diff
@@ -973,7 +973,12 @@ Query_route_to Query::Compile(CompiledQuery *compiled_query, Query_block *select
     // according the meaning of `part`, which describs in JOIN::optimize
     if (!sl->join) {
-      TIANMU_LOG(LogCtl_Level::ERROR, "sl->join is nil!!!!");
+      auto thd = current_txn_->Thd();
```
Why the `join` is null at here, we think it should be already created at sql.  Here, we can fix the crash, but what's the root cause of `join` is not created by now.

lujiashun replied:
In StoneDB5.6, the subquery join is created in the backtrace
```
#0  JOIN::JOIN (this=0x556c083d9b16 <Sql_alloc::operator new(unsigned long)+28>, thd_arg=0x7ff3600e0cc0, fields_arg=...,
    select_options_arg=140683265314008, result_arg=0x440) at /data/codebase/stonedb56/stonedb/sql/sql_optimizer.h:401
#1  0x0000556c086e48cd in subselect_single_select_engine::prepare (this=0x7fee98007050) at /data/codebase/stonedb56/stonedb/sql/item_subselect.cc:2625
#2  0x0000556c086dd06c in Item_subselect::fix_fields (this=0x7fee98006ee0, thd=0x556c0a5791f0, ref=0x7fee98007128)
    at /data/codebase/stonedb56/stonedb/sql/item_subselect.cc:381
#3  0x0000556c086e4003 in Item_in_subselect::fix_fields (this=0x7fee98006ee0, thd_arg=0x556c0a5791f0, ref=0x7fee98007128)
    at /data/codebase/stonedb56/stonedb/sql/item_subselect.cc:2391
#4  0x0000556c0843aa75 in Item_func::fix_fields (this=0x7fee98007088, thd=0x556c0a5791f0, ref=0x7fee9801aa00)
    at /data/codebase/stonedb56/stonedb/sql/item_func.cc:203
#5  0x0000556c08512462 in setup_conds (thd=0x556c0a5791f0, tables=0x7fee980054e8, leaves=0x7fee980054e8, conds=0x7fee9801aa00)
    at /data/codebase/stonedb56/stonedb/sql/sql_base.cc:8872
#6  0x0000556c085a0fdb in setup_without_group (thd=0x556c0a5791f0, ref_pointer_array=..., tables=0x7fee980054e8, leaves=0x7fee980054e8, fields=...,
    all_fields=..., conds=0x7fee9801aa00, order=0x0, group=0x0, hidden_group_field_count=0x7fee9801a8e4, hidden_order_field_count=0x7ff3600e1200)
    at /data/codebase/stonedb56/stonedb/sql/sql_resolver.cc:952
#7  0x0000556c0859e750 in JOIN::prepare (this=0x7fee9801a6c0, tables_init=0x7fee980054e8, wild_num=1, conds_init=0x7fee98007088, og_num=0,
    order_init=0x0, group_init=0x0, having_init=0x0, select_lex_arg=0x556c0a57b6b8, unit_arg=0x556c0a57b070)
    at /data/codebase/stonedb56/stonedb/sql/sql_resolver.cc:173
#8  0x0000556c08c35c14 in stonedb::core::optimize_select (thd=0x556c0a5791f0, tables=0x7fee980054e8, wild_num=1, fields=..., conds=0x7fee98007088,
    og_num=0, order=0x0, group=0x0, having=0x0, select_options=2147748608, result=0x7fee980071f8, unit=0x556c0a57b070, select_lex=0x556c0a57b6b8,
    optimize_after_sdb=@0x7ff3600e15fc: 0, free_join=@0x7ff3600e15f8: 1) at /data/codebase/stonedb56/stonedb/storage/stonedb/core/engine_execute.cpp:332
#9  0x0000556c08c35503 in stonedb::core::Engine::HandleSelect (this=0x556c0a5d3c90, thd=0x556c0a5791f0, lex=0x556c0a57afb0,
    result=@0x7ff3600e1600: 0x7fee980071f8, setup_tables_done_option=0, res=@0x7ff3600e15f4: 0, optimize_after_sdb=@0x7ff3600e15fc: 0,
    sdb_free_join=@0x7ff3600e15f8: 1, with_insert=0) at /data/codebase/stonedb56/stonedb/storage/stonedb/core/engine_execute.cpp:223
#10 0x0000556c08d13bd8 in stonedb::dbhandler::SDB_HandleSelect (thd=0x556c0a5791f0, lex=0x556c0a57afb0, result=@0x7ff3600e1600: 0x7fee980071f8,
    setup_tables_done_option=0, res=@0x7ff3600e15f4: 0, optimize_after_sdb=@0x7ff3600e15fc: 0, sdb_free_join=@0x7ff3600e15f8: 1, with_insert=0)
    at /data/codebase/stonedb56/stonedb/storage/stonedb/handler/ha_rcengine.cpp:81
#11 0x0000556c0857c0e8 in execute_sqlcom_select (thd=0x556c0a5791f0, all_tables=0x7fee980054e8)
    at /data/codebase/stonedb56/stonedb/sql/sql_parse.cc:5170
#12 0x0000556c08574ae0 in mysql_execute_command (thd=0x556c0a5791f0) at /data/codebase/stonedb56/stonedb/sql/sql_parse.cc:2663
#13 0x0000556c0857eed2 in mysql_parse (thd=0x556c0a5791f0,
    rawbuf=0x7fee98005290 "select * from t1 where val not in (select * from t2 where val2 >= 10)", length=69, parser_state=0x7ff3600e27a0)
    at /data/codebase/stonedb56/stonedb/sql/sql_parse.cc:6428
#14 0x0000556c08571959 in dispatch_command (command=COM_QUERY, thd=0x556c0a5791f0,
    packet=0x556c0d0754e1 "select * from t1 where val not in (select * from t2 where val2 >= 10)", packet_length=69)
    at /data/codebase/stonedb56/stonedb/sql/sql_parse.cc:1342
#15 0x0000556c08570997 in do_command (thd=0x556c0a5791f0) at /data/codebase/stonedb56/stonedb/sql/sql_parse.cc:1039

```

lujiashun replied:
test case for debug this issue
```
use test;
create table t1(val int)  ENGINE=tianmu;
create table t2(val2 int) ENGINE=tianmu;
select * from t1 where val not in (select * from t2 where val2 >= 10);
create table t3(val int)  ENGINE=innodb;
create table t4(val2 int) ENGINE=innodb;
select * from t3 where val not in (select * from t4 where val2 >= 10);
```

lujiashun replied:
In StoneDB8.0, the join is created here
```
(gdb) bt
#0  JOIN::JOIN (this=0x55b40f70b708 <operator new(unsigned long, MEM_ROOT*, std::nothrow_t const&)+43>, thd_arg=0x7f7e284072c0, select=0x7f78e00f2b70)
    at /data/codebase/stonedb80/stonedb/sql/sql_optimizer.cc:158
#1  0x000055b40f92ba3e in Query_block::optimize (this=0x7f78e014b828, thd=0x7f78e0001040, finalize_access_paths=true)
    at /data/codebase/stonedb80/stonedb/sql/sql_select.cc:1819
#2  0x000055b40f9e291d in Query_expression::optimize (this=0x7f78e014b078, thd=0x7f78e0001040, materialize_destination=0x0, create_iterators=false,
    finalize_access_paths=true) at /data/codebase/stonedb80/stonedb/sql/sql_union.cc:701
#3  0x000055b40f92bb31 in Query_block::optimize (this=0x7f78e00ecdf0, thd=0x7f78e0001040, finalize_access_paths=true)
    at /data/codebase/stonedb80/stonedb/sql/sql_select.cc:1839
#4  0x000055b40f9e291d in Query_expression::optimize (this=0x7f78e00ec640, thd=0x7f78e0001040, materialize_destination=0x0, create_iterators=true,
    finalize_access_paths=true) at /data/codebase/stonedb80/stonedb/sql/sql_union.cc:701
#5  0x000055b40f9291f2 in Sql_cmd_dml::execute_inner (this=0x7f78e014cd58, thd=0x7f78e0001040) at /data/codebase/stonedb80/stonedb/sql/sql_select.cc:783
#6  0x000055b40f928744 in Sql_cmd_dml::execute (this=0x7f78e014cd58, thd=0x7f78e0001040) at /data/codebase/stonedb80/stonedb/sql/sql_select.cc:597
#7  0x000055b40f89f5cf in mysql_execute_command (thd=0x7f78e0001040, first_level=true) at /data/codebase/stonedb80/stonedb/sql/sql_parse.cc:4604
#8  0x000055b40f8a18ba in dispatch_sql_command (thd=0x7f78e0001040, parser_state=0x7f7e28408b90)
    at /data/codebase/stonedb80/stonedb/sql/sql_parse.cc:5239
#9  0x000055b40f897339 in dispatch_command (thd=0x7f78e0001040, com_data=0x7f7e284094e0, command=COM_QUERY)
    at /data/codebase/stonedb80/stonedb/sql/sql_parse.cc:1959
#10 0x000055b40f8953b0 in do_command (thd=0x7f78e0001040) at /data/codebase/stonedb80/stonedb/sql/sql_parse.cc:1362
#11 0x000055b40fad1ce9 in handle_connection (arg=0x55b41904a300)
    at /data/codebase/stonedb80/stonedb/sql/conn_handler/connection_handler_per_thread.cc:302
#12 0x000055b4119f3bce in pfs_spawn_thread (arg=0x55b41b0b4cc0) at /data/codebase/stonedb80/stonedb/storage/perfschema/pfs.cc:2942
#13 0x00007f7e5371d609 in start_thread (arg=<optimized out>) at pthread_create.c:477
#14 0x00007f7e52628133 in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:95

```

RingsC reviewed (approved):
LGTM

DandreChen reviewed (approved):
LGTM

Commit:
fix(tianmu): fix query syntax (WHERE NOT IN ) is not supported. (#767)

[summary]
1 'sub_select' item is the left argument,not the right argument of 'item_func';
2 5.7 use 'item_func_not' class to represent the meaning of the syntax "not", while 8.0 use a variable 'value_transform';
3 porting new join code from 5.7 for this case；
3.1 solve throwing exception ”select * from t1 where val not in (select * from t2 where val2 >= 10);"
4 porting ChooseJoinAlgorithm from 5.7 for this case;
4.1 solve incorrect result set for some sql in the mtr（which one is forgotton...)
5 porting ResetToTemplate modification from 5.7 for this case;
5.1 solve incorrect result set "select * from t1 where val not in (select * from t2 where val2 > t1.val);"
